backend/alternative_flights/alt_flight_search.py

```diff
@@ -169,22 +169,41 @@ def apply_discounts(options: List[FlightOption], baseline_duration: float):
         option.discount = min(discount, option.total_price * 0.7)  # Cap discount at 70%
         option.total_price -= option.discount
 
-def calculate_ranking_scores(options: List[FlightOption], preferred_airline_code: str) -> List[FlightOption]:
+def calculate_ranking_scores(options: List[FlightOption], preferred_airline_code: str, airline_usage: List[Dict]) -> List[FlightOption]:
     prices = np.array([opt.total_price for opt in options])
     durations = np.array([opt.total_duration for opt in options])
     connections = np.array([len(opt.flights) - 1 for opt in options])
-    airline_matches = np.array([1 if any(flight.airline_code == preferred_airline_code for flight in opt.flights) else 0 for opt in options])
 
     price_scores = 1 - (prices - np.min(prices)) / (np.max(prices) - np.min(prices))
     speed_scores = 1 - (durations - np.min(durations)) / (np.max(durations) - np.min(durations))
     directness_scores = np.where(connections == 0, 2, 1 - connections / np.max(connections))
 
-    overall_scores = (
-        0.35 * price_scores +
-        0.25 * speed_scores +
-        0.30 * directness_scores +
-        0.10 * airline_matches
-    )
+    preferred_airline_scores = np.array([
+        1 if any(flight.airline_code == preferred_airline_code for flight in opt.flights) else 0
+        for opt in options
+    ])
+
+    usage_dict = {usage['airlineCode']: usage['count'] for usage in airline_usage}
+    max_usage = max(usage_dict.values(), default=0)
+    if max_usage > 0:
+        historical_usage_scores = np.array([
+            max((usage_dict.get(flight.airline_code, 0) / max_usage for flight in opt.flights), default=0)
+            for opt in options
+        ])
+        overall_scores = (
+            0.25 * price_scores +
+            0.20 * speed_scores +
+            0.20 * directness_scores +
+            0.20 * preferred_airline_scores +
+            0.15 * historical_usage_scores
```

Comment:
Just to make sure I'm understanding correctly. If I have traveled with Delta 50 times, but 0 with other airlines, flights including Delta will get +0.15 score. Is this correct?

Another example:
If I traveled with Delta 10 times, United 10 times and Alaska 10 times; a recommendation including Delta and United in connected flights but not Alaska would score +0.1.

Is this accurate?

Reply:
1. Yes 
2. Did you mean +.15 for the second scenario? If so, yes.

Reply:
Ok for #2, I would appreciate a more detailed explanation to understand how this historical scores feature work.

The more you have traveled with an airline, the higher the score (right?).

1. If I traveled 50 times with Delta and 50 times with United; what would be the "airline preference" score if I take a single flight trip with Delta? Would it be 0.15 or 0.15/2 ?

2. How is it supposed to work with connected flights?

Reply:
Yes, The more we have traveled with an airline, the higher the score. One important thing to note for both question is –  the airline usage score is based on the maximum usage count for any airline in the trip.

1. So far a single flight trip with Delta, 

- If you have traveled 50 times with Delta and 50 times with United, the maximum usage count is 50.

- For a single flight trip with Delta, the historical usage score would be:

           Historical Usage Score = Delta Usage Count / Max Usage Count = 50 / 50 = 1

- Since the historical usage score contributes 0.15 to the overall score, the contribution would be:

0.15 * 1 = 0.15

2. For connected flights, it works the same way. 

A trip with both Delta and United would also get a 0.15 contribution. That’s because the algorithm considers the maximum normalized score among all airlines in the itinerary.

- Both Delta and United have a normalized score of 1 (50/50)
- The maximum of these scores is still 1
- So the historical usage score remains 1 * 0.15 = 0.15, just like with a single flight

This ensures that connected flights aren't given an unfair advantage over direct flights in terms of historical usage scoring.